packages/space-opera/src/components/materials_panel/materials_panel.ts

```diff
@@ -295,358 +309,257 @@ export class MaterialPanel extends ConnectedLitElement {
   }
 
-  onBaseColorTextureUpload(event: CustomEvent) {
-    if (this.selectedMaterialId === undefined) {
-      throw new Error('No material selected');
+  onRoughnessChange() {
+    this.getMaterial().pbrMetallicRoughness.setRoughnessFactor(
+        this.selectedRoughnessFactor);
+    reduxStore.dispatch(dispatchModelDirty());
+  }
+
+  onMetallicChange() {
+    this.getMaterial().pbrMetallicRoughness.setMetallicFactor(
+        this.selectedMetallicFactor);
+    reduxStore.dispatch(dispatchModelDirty());
+  }
+
+  onDoubleSidedChange(event: Event) {
+    this.updateDoubleSided((event.target as HTMLInputElement).checked);
+  }
+
+  updateDoubleSided(value: boolean) {
+    this.getMaterial().setDoubleSided(value);
+    reduxStore.dispatch(dispatchModelDirty());
+  }
+
+  onTextureChange(textureId: string|undefined, textureInfo: TextureInfo) {
+    const texture = textureId != null ?
+        this.thumbnailsById.get(textureId)?.texture :
+        undefined;
+    textureInfo.setTexture(texture as any ?? null);
+    reduxStore.dispatch(dispatchModelDirty());
+  }
+
+  async onTextureUpload(uri: string, textureInfo: TextureInfo) {
+    if (this.thumbnailsById.has(uri)) {
+      console.log('URL collision! Texture not updated.');
+      return;
+    }
+    const texture = await getModelViewer()?.createTexture(uri);
+    if (texture == null) {
+      return;
+    }
+    textureInfo.setTexture(texture);
+    const id = await pushThumbnail(this.thumbnailsById, textureInfo);
+    // Trigger async panel update / render
+    this.thumbnailsById = new Map(this.thumbnailsById);
+    if (id != null) {
+      this.thumbnailIds.push(id);
+      // Trigger async texture_picker update / render
+      this.thumbnailUrls = [...this.thumbnailUrls];
+      this.thumbnailUrls.push(this.thumbnailsById.get(id)!.objectUrl);
     }
+    reduxStore.dispatch(dispatchModelDirty());
+    this.dispatchEvent(new CustomEvent('texture-upload-complete'));
+  }
+
+  onBaseColorTextureChange() {
+    this.onTextureChange(
+        this.selectedBaseColorTextureId,
+        this.getMaterial().pbrMetallicRoughness.baseColorTexture);
+  }
 
-    const id = this.selectedMaterialId;
-    const uri = event.detail;
-    reduxStore.dispatch(dispatchAddBaseColorTexture(
-        getEditsMaterials(reduxStore.getState()),
-        getEditsTextures(reduxStore.getState()),
-        {id, uri}));
+  onBaseColorTextureUpload(event: CustomEvent) {
+    this.onTextureUpload(
+        event.detail, this.getMaterial().pbrMetallicRoughness.baseColorTexture);
   }
 
   onMetallicRoughnessTextureChange() {
-    if (this.selectedMaterialId === undefined) {
-      throw new Error('No material selected');
-    }
-    const id = this.selectedMaterialId;
-    const textureId = this.selectedMetallicRoughnessTextureId;
-    reduxStore.dispatch(dispatchMetallicRoughnessTexture(
-        getEditsMaterials(reduxStore.getState()), {id, textureId}));
+    this.onTextureChange(
```

Comment:
This change is nice, is there a way to get rid of all the bespoke OnThisTextureChange() methods entirely and just use this instead?

Reply:
Yeah, I was thinking the same thing. I'll give it a shot.

Reply:
Hmm, it looks like because of the way the HTML is interpreted, that's not so easy. Probably why they did it this way in the first place.

Reply:
Yeah I thought maybe something like that, overall this is much better